core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -179,7 +180,11 @@ class HistoryServer(
   }
```

Comment:
OK, let me ask a different question -- when would you ever want to call getApplicationInfoList, given this problem? the other usages seem to just pick out a single element, so it's wasteful for the same reason. It just seems weird to have two methods for this.

The existing method already only promises an Iterator which seems like the right thing to do. The problem is that it gets there by materializing a whole collection. Why is that unavoidable -- you can map() an iterator's elements without any evaluation of its elements or needing to first materialize the collection. 

That's why I think getApplicationInfo is really what needs to change to provide something else like an Iterator, and that ends up being pretty easy -- FsHistoryProvider just has to give an iterator over applications.